tests/offline/test_acceptance.py

```diff
@@ -0,0 +1,170 @@
+from dsi.types.result import ResultAcceptance
+
+
+@pytest.fixture
+def experiment():
+    config = ConfigAcceptanteRate(
+        model="target_model", dataset="dataset", draft_model="draft_model"
+    )
+    gen_config = ConfigGen()
+    draft_gen_config = ConfigGen()
+    return ExperimentAcceptanceRate(config, gen_config, draft_gen_config)
+
+
+@pytest.fixture
+def mock_dependencies():
+    with patch(
+        "transformers.AutoModelForCausalLM.from_pretrained"
+    ) as mock_model, patch(
+        "transformers.AutoTokenizer.from_pretrained"
+    ) as mock_tokenizer, patch.object(
+        ExperimentAcceptanceRate, "_get_random_prompted_examples"
+    ) as mock_examples:
+        mock_model.return_value = MagicMock()
+        mock_model.return_value.device = torch.device("cpu")
+        mock_tokenizer.return_value = MagicMock()
+        mock_examples.return_value = ["example1"]
+        yield mock_model, mock_tokenizer, mock_examples
+
+
+def test_are_tokenizers_same_identical(experiment):
+    tokenizer1 = AutoTokenizer.from_pretrained("double7/vicuna-68m")
+    tokenizer2 = AutoTokenizer.from_pretrained("double7/vicuna-68m")
+    assert experiment._are_tokenizers_same(tokenizer1, tokenizer2)
+
+
+def test_are_tokenizers_same_diff_config(experiment):
+    tokenizer1 = MagicMock()
+    tokenizer2 = MagicMock()
+    tokenizer1.config = {"model_type": "bigcode/starcoder"}
+    tokenizer2.config = {"model_type": "double7/vicuna-68m"}
+    assert not experiment._are_tokenizers_same(tokenizer1, tokenizer2)
+
+
+def test_are_tokenizers_same_diff_vocab(experiment):
+    tokenizer1 = MagicMock()
+    tokenizer2 = MagicMock()
+    tokenizer1.get_vocab.return_value = {"hello": 1, "world": 2}
+    tokenizer2.get_vocab.return_value = {"hello": 1, "python": 3}
+    assert not experiment._are_tokenizers_same(tokenizer1, tokenizer2)
+
+
+def test_are_tokenizers_same_diff_special_tokens(experiment):
+    tokenizer1 = MagicMock()
+    tokenizer2 = MagicMock()
+    tokenizer1.eos_token_id = 1
+    tokenizer2.eos_token_id = 2
+    tokenizer1.pad_token_id = 0
+    tokenizer2.pad_token_id = 0
+    tokenizer1.bos_token_id = -1
+    tokenizer2.bos_token_id = -1
+    tokenizer1.unk_token_id = 3
+    tokenizer2.unk_token_id = 3
+    assert not experiment._are_tokenizers_same(tokenizer1, tokenizer2)
+
+
+def test_single_repeat_all_match(experiment, mock_dependencies):
+    mock_model, _, _ = mock_dependencies
+    # Mock the generate method to simulate target and draft models
+    # producing the same output
+    mock_model.return_value.generate.side_effect = [
+        torch.tensor([[0, 1, 2, 3]]),
+        torch.tensor([[0]]),
+        torch.tensor([[0, 1]]),
+        torch.tensor([[0, 1, 2]]),
+        torch.tensor([[0, 1, 2, 3]]),
+    ]
+    result = experiment._single_repeat()
+    assert isinstance(result, ResultAcceptance)
+    # Since all tokens match, acceptance rate should be 1
+    assert result.acceptance_rate[0] == 0.8
```

Comment:
It seems like a discrepancy. The test passes, suggesting that there is a bug.

Reply:
@keyboardAnt 
for the paper, we agreed to ignore the last matching window for each dataset input, since generation can stop because we reach EOS or max length, and not because draft model bad prediction.
Not sure we wrote it in the paper (please correct me). 
According to the paper formula, AR is 0.8.

Reply:
@jmamou
Even when ignoring the test's 4th (and last) match, why the returned acceptance rate is < 1? The first 3 tokens match in 100%.

Reply:
@keyboardAnt 
For the paper, we agreed to ignore the last matching window, not the last matching token.
In real scenarios, we will never get AR=1 since n cannot be infinite.

Reply:
@jmamou 
1. What is the definition of acceptance rate for such windows? How is it different from the acceptance rate for single tokens? (For single tokens, the acceptance rate is the probability of accepting the drafter's next token prediction using an exact match or Miao's rejection sampling algorithm, assuming iid target tokens)
2. Why is ignoring the last matching token counted as a mismatch in `test_single_repeat_all_match`? Please correct me if I'm wrong here. In the test, the target [outputs](https://github.com/keyboardAnt/distributed-speculative-inference/pull/34/files/bc3f79c9d2a7bfad4b720f3e579eb3afdb98df93#diff-bd2ff502bfd91f8011176875cc43e1ea16f580949e10186f3b57959d6b37c810R80) the sequence of four input ids `[0, 1, 2, 3]`. Then, the drafter correctly [predicts](https://github.com/keyboardAnt/distributed-speculative-inference/pull/34/files/bc3f79c9d2a7bfad4b720f3e579eb3afdb98df93#diff-bd2ff502bfd91f8011176875cc43e1ea16f580949e10186f3b57959d6b37c810R81-R84) all four. However, the calculated acceptance rate [equals](https://github.com/keyboardAnt/distributed-speculative-inference/pull/34/files/bc3f79c9d2a7bfad4b720f3e579eb3afdb98df93#diff-bd2ff502bfd91f8011176875cc43e1ea16f580949e10186f3b57959d6b37c810R89) 0.8. Why the acceptance rate isn't 1?

Reply:
As we discussed today over the phone, the acceptance rate has only one definition:
> The acceptance rate is the probability of accepting the drafter's next token prediction using an exact match or Miao's rejection sampling algorithm, assuming iid target tokens.

We estimate the acceptance rate as follows. Let `A` be the sum of the number of drafts accepted over all iterations except the last iteration over all the examples in all datasets. Denote the total number of such iterations by `N`. If `A` equals the number of examples, the drafter predicted 100% of the example in a single iteration. In other words, we accepted all draft tokens without any rejections. In that case, we can **raise a warning and return 1 _(recommended)_** or raise an exception. Otherwise, we calculate the average number of accepted drafts `n := A/N` (dividing the sum by the total number of iterations). The estimated acceptance rate is then `1 - 1 / (1 + n)` as mentioned in the paper. The issue with `test_single_repeat_all_match` is that `1 - 1 / (1 + n) != 1` for any `n`. 

In practice, drafters may have a 100% acceptance rate. For example, an instance of the target running on faster hardware as a drafter, with both the target and drafter sampled greedily. But in such cases the perfect acceptance rate is guaranteed, and I do not see a reason to test it using our experiment. On the contrary, in most practical settings, the acceptance rate is < 1, and estimating it as = 1 means we haven't reached the false discovery rate. In other words, we need to consider more examples. Also, it might indicate a bug. So, this is why I think we should raise a warning or exception.

@jmamou, lmkwyt